8351347: HttpClient Improve logging of response headers

vy opened the pull request:
Includes request method, request URI, response status code, and HTTP/2 stream ID while logging response headers in the HTTP Client.

### Demonstration

Snippets from running JTreg against `test/jdk/java/net/httpclient/HeadTest.java`:

**Before:**
```
INFO: HEADERS: REQUEST HEADERS:
HEAD /transfer/ HTTP/1.1
Content-Length: 0
Host: 127.0.0.1:43647
...
INFO: HEADERS: RESPONSE HEADERS:
    connection: Upgrade
    upgrade: h2c
...
INFO: HEADERS: RESPONSE HEADERS:
    :status: 304
    content-length: 300
...
INFO: HEADERS: HEADERS FRAME (streamid=1):
    :authority: 127.0.0.1:50611
    :method: GET
    :path: /
```

**After:** 
```
INFO: HEADERS: REQUEST HEADERS:
  HEAD /transfer/ HTTP/1.1
    Content-Length: 0
    Host: 127.0.0.1:43647
...
INFO: HEADERS: RESPONSE HEADERS:
  GET http://127.0.0.1:48497/ 101
    connection: Upgrade
    upgrade: h2c
...
INFO: HEADERS: RESPONSE HEADERS (streamid=1):
  GET http://127.0.0.1:48497/ 304
    :status: 304
    content-length: 300
...
INFO: HEADERS: HEADERS FRAME (streamid=1):
  GET https://127.0.0.1:50611/
    :authority: 127.0.0.1:50611
    :method: GET
    :path: /
```

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8351347](https://bugs.openjdk.org/browse/JDK-8351347): HttpClient Improve logging of response headers (**Bug** - P4)


### Reviewers
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**)
 * [Michael McMahon](https://openjdk.org/census#michaelm) (@Michael-Mc-Mahon - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/25201/head:pull/25201` \
`$ git checkout pull/25201`

Update a local copy of the PR: \
`$ git checkout pull/25201` \
`$ git pull https://git.openjdk.org/jdk.git pull/25201/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 25201`

View PR using the GUI difftool: \
`$ git pr show -t 25201`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/25201.diff">https://git.openjdk.org/jdk/pull/25201.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/25201#issuecomment-2875505694)
</details>

dfuch commented:
Very nice!

It would be good to add the `(streamId=<streamId>)` at line 676 too in Stream.java (Trailing Headers),  and to add the whole line (GET <uri> <status-code>) at line 1782 (push response headers).

vy commented:
@dfuch, I've updated all usages of `Log::logHeaders`. Two particular spots of importance:
1. Updated `Http1Request::logHeaders`
2. `HEADERS FRAME` logs don't contain the response code for those are outgoing

dfuch commented on src/java.net.http/share/classes/jdk/internal/net/http/Http1Request.java:
```diff
@@ -81,15 +81,22 @@ class Http1Request {
+                // First line contains `GET /foo HTTP/1.1`.
+                // Convert it to look like other `Log.logHeaders()` outputs.
+                if (firstLine[0]) {
+                    int protocolFieldPrefixIndex = line.lastIndexOf(' ');
+                    assert protocolFieldPrefixIndex > 0;
+                    sb.append("  ").append(line, 0, protocolFieldPrefixIndex).append('\n');
```
I'd rather print the whole line here - including HTTP/1.1.
Nice to have it indented differently though.

vy replied:
Implemented in 7b958db6f4.

dfuch commented on src/java.net.http/share/classes/jdk/internal/net/http/Stream.java:
```diff
@@ -673,7 +674,8 @@ request, exchange, responseHeaders, connection(),
             if (Log.headers()) {
-                StringBuilder sb = new StringBuilder("TRAILING HEADERS:\n");
+                StringBuilder sb = new StringBuilder("TRAILING HEADERS (streamid=%d):\n".formatted(streamid));
+                sb.append("  %s %s %d\n".formatted(request.method(), request.uri(), responseCode));
```
These are trailing headers - printing URI and response code would be confusing here. Only printing streamId should be fine.

```suggestion
```

dfuch commented on src/java.net.http/share/classes/jdk/internal/net/http/Stream.java:
```diff
@@ -1770,8 +1772,8 @@ pushReq, exchange, responseHeaders, connection(),
                 if (Log.headers()) {
-                    StringBuilder sb = new StringBuilder("RESPONSE HEADERS");
-                    sb.append(" (streamid=").append(streamid).append("):\n");
+                    StringBuilder sb = new StringBuilder("RESPONSE HEADERS (streamid=%d):\n".formatted(streamid));
```
```suggestion
                    StringBuilder sb = new StringBuilder("RESPONSE HEADERS (streamid=%s):\n".formatted(streamid));
```

dfuch commented on src/java.net.http/share/classes/jdk/internal/net/http/Stream.java:
```diff
@@ -673,7 +674,8 @@ request, exchange, responseHeaders, connection(),
         } else {
             if (Log.headers()) {
-                StringBuilder sb = new StringBuilder("TRAILING HEADERS:\n");
+                StringBuilder sb = new StringBuilder("TRAILING HEADERS (streamid=%d):\n".formatted(streamid));
```
```suggestion
                StringBuilder sb = new StringBuilder("TRAILING HEADERS (streamid=%s):\n".formatted(streamid));
```

dfuch commented on src/java.net.http/share/classes/jdk/internal/net/http/Stream.java:
```diff
@@ -662,7 +662,8 @@ request, exchange, responseHeaders, connection(),
             if (Log.headers()) {
-                StringBuilder sb = new StringBuilder("RESPONSE HEADERS:\n");
+                StringBuilder sb = new StringBuilder("RESPONSE HEADERS (streamid=%d):\n".formatted(streamid));
+                sb.append("  %s %s %d\n".formatted(request.method(), request.uri(), responseCode));
```
```suggestion
                sb.append("  %s %s %s\n".formatted(request.method(), request.uri(), responseCode));
```

dfuch commented on src/java.net.http/share/classes/jdk/internal/net/http/Http2Connection.java:
```diff
@@ -1685,8 +1685,8 @@ private List<ByteBuffer> encodeHeadersImpl(int bufferSize, HttpHeaders... header
         if (Log.headers()) {
-            StringBuilder sb = new StringBuilder("HEADERS FRAME (stream=");
-            sb.append(stream.streamid).append(")\n");
+            StringBuilder sb = new StringBuilder("HEADERS FRAME (streamid=%d):\n".formatted(stream.streamid));
```
```suggestion
            StringBuilder sb = new StringBuilder("HEADERS FRAME (streamid=%s):\n".formatted(stream.streamid));
```

dfuch commented on src/java.net.http/share/classes/jdk/internal/net/http/Http1Response.java:
```diff
@@ -224,6 +224,7 @@ public CompletableFuture<Response> readHeadersAsync(Executor executor) {
 
                 if (Log.headers()) {
                     StringBuilder sb = new StringBuilder("RESPONSE HEADERS:\n");
+                    sb.append("  %s %s %d\n".formatted(request.method(), request.uri(), responseCode));
```
```suggestion
                    sb.append("  %s %s %s\n".formatted(request.method(), request.uri(), responseCode));
```

dfuch commented on src/java.net.http/share/classes/jdk/internal/net/http/Stream.java:
```diff
@@ -1770,8 +1772,8 @@ pushReq, exchange, responseHeaders, connection(),
-                    StringBuilder sb = new StringBuilder("RESPONSE HEADERS");
-                    sb.append(" (streamid=").append(streamid).append("):\n");
+                    StringBuilder sb = new StringBuilder("RESPONSE HEADERS (streamid=%d):\n".formatted(streamid));
+                    sb.append("  %s %s %d\n".formatted(request.method(), request.uri(), responseCode));
```
```suggestion
                    sb.append("  %s %s %s\n".formatted(request.method(), request.uri(), responseCode));
```

dfuch commented on src/java.net.http/share/classes/jdk/internal/net/http/Stream.java:
```diff
@@ -1782,8 +1784,8 @@ pushReq, exchange, responseHeaders, connection(),
                 if (Log.headers()) {
-                    StringBuilder sb = new StringBuilder("TRAILING HEADERS");
-                    sb.append(" (streamid=").append(streamid).append("):\n");
+                    StringBuilder sb = new StringBuilder("TRAILING HEADERS (streamid=%d):\n".formatted(streamid));
+                    sb.append("  %s %s %d\n".formatted(request.method(), request.uri(), responseCode));
```
```suggestion
                    StringBuilder sb = new StringBuilder("TRAILING HEADERS (streamid=%s):\n".formatted(streamid));
                    sb.append("  %s %s %s\n".formatted(request.method(), request.uri(), responseCode));
```

dfuch reviewed (commented):
Some suggested changes. I'd rather not localize streamId and status code - so I'd prefer we use %s.

vy commented on src/java.net.http/share/classes/jdk/internal/net/http/Http1Request.java:
```diff
@@ -81,15 +81,22 @@ class Http1Request {
+                // First line contains `GET /foo HTTP/1.1`.
+                // Convert it to look like other `Log.logHeaders()` outputs.
+                if (firstLine[0]) {
+                    int protocolFieldPrefixIndex = line.lastIndexOf(' ');
+                    assert protocolFieldPrefixIndex > 0;
+                    sb.append("  ").append(line, 0, protocolFieldPrefixIndex).append('\n');
```
```suggestion
                    sb.append("  ").append(line).append('\n');
```

vy commented on src/java.net.http/share/classes/jdk/internal/net/http/Stream.java:
```diff
@@ -662,7 +662,8 @@ request, exchange, responseHeaders, connection(),
 
             if (Log.headers()) {
-                StringBuilder sb = new StringBuilder("RESPONSE HEADERS:\n");
+                StringBuilder sb = new StringBuilder("RESPONSE HEADERS (streamid=%d):\n".formatted(streamid));
```
```suggestion
                StringBuilder sb = new StringBuilder("RESPONSE HEADERS (streamid=%s):\n".formatted(streamid));
```

vy commented:
@dfuch, thanks for the `%d`-vs-`%s` tip, agreed with your rationale. :100:

Applied requested changes and updated the PR description.

Michael-Mc-Mahon reviewed (approved):
Nice work

vy commented:
Successful `tier1,2` results are attached to the ticket.

/integrate

dfuch commented:
/sponsor
